WebView/app/src/main/java/com/android/samples/webviewdemo/MainActivity.kt

```diff
@@ -43,26 +50,71 @@ class MainActivity : AppCompatActivity() {
-    /** Instantiate the interface and set the context  */
-    class WebAppInterface(private val mContext: Context) {
-        /** Send a message from the web page  */
-        @JavascriptInterface
-        fun sendMessage(message: String) {
-            val sendIntent: Intent = Intent().apply {
-                action = Intent.ACTION_SEND
-                putExtra(Intent.EXTRA_TEXT, message)
-                type = "text/plain"
-            }
-            val shareIntent = Intent.createChooser(sendIntent, null)
-            startActivity(mContext, shareIntent, null)
+    /**
+     * This method checks to see if WebMessageListener is supported. This method is prefered
+     * for security reasons. If this feature is supported, then it is used to create a
+     * Javascript object to be injected into every iframe. If it is not supported then
+     * then the app will fall back to JavascriptInterface.Here, the custom object that is inserted
+     * using JavascriptInterface is setup to mimic the object that would be created by WebMessageListener
+     * if it were available. This is so that the that Javascript will not have to behave
+     * any differently whether it is passed an object by WebMessageListener or JavascriptInterface.
+     * @param webview the component that is
+     * @param mContext the amount of incoming damage
+     * @param jsObjName the name that will be given to the Javascript objects created by either
+     *        WebMessageListener or JavascriptInterface
+     * @param allowedOriginRules a set of rules, a frame must match an Origin in this set to have
+     *        the JS object injected into it
+     * @param onMessageReceived a callback which invokes native android sharing
+     * @return void
+     */
+    private fun createJsObject(
+        webview: WebView,
+        jsObjName: String,
+        allowedOriginRules: Set<String>,
+        onMessageReceived: (message: String) -> Unit
+    ) {
+        if (WebViewFeature.isFeatureSupported(WebViewFeature.WEB_MESSAGE_LISTENER)) {
+            WebViewCompat.addWebMessageListener(
+                webview, jsObjName, allowedOriginRules,
+                object : WebViewCompat.WebMessageListener {
+                    override fun onPostMessage(
+                        webview: WebView,
+                        message: WebMessageCompat,
+                        sourceOrigin: Uri,
+                        isMainFrame: Boolean,
+                        replyProxy: JavaScriptReplyProxy
+                    ) {
+                        onMessageReceived(message.data!!)
+                    }
+                })
+        } else {
+            webview.addJavascriptInterface(object {
+                @JavascriptInterface
+                fun postMessage(message: String) {
+                    onMessageReceived(message)
+                }
+            }, jsObjName)
         }
+    }
 
+    // Invokes native android sharing
+    private val onMessageReceived = { message: String ->
```

Comment:
I think this would read better by either (a) moving this directly into the `createJsObject` call site below (that's the only spot it's used) or (b) making this a named method of the Activity (ex. `createShareIntent(String message)`) and passing a method reference into `createJsObject`.

Reply:
For now I've made it a method on the Activity. I think that helps with the goal of keeping `createJsObject` as light as possible.